### ISSUE TITLE
from __future__ import annotations breaks profiler's handling of dataclasses

### ISSUE THREAD

**Issue by cdd13971-1229-447b-8f55-6657365cb898** [python/cpython#88580]
BPO | [44414](https://bugs.python.org/issue44414)
--- | :---
Nosy | @JelleZijlstra, @wilcoxjay

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-06-13.21:20:30.284>
labels = ['type-bug', 'library', '3.9']
title = "from __future__ import annotations breaks profiler's handling of dataclasses"
updated_at = <Date 2021-06-14.04:51:07.905>
user = 'https://github.com/wilcoxjay'
```

bugs.python.org fields:
```python
activity = <Date 2021-06-14.04:51:07.905>
actor = 'JelleZijlstra'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-06-13.21:20:30.284>
creator = 'wilcoxjay'
dependencies = []
files = []
hgrepos = []
issue_num = 44414
keywords = []
message_count = 1.0
messages = ['395762']
nosy_count = 2.0
nosy_names = ['JelleZijlstra', 'wilcoxjay']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue44414'
versions = ['Python 3.9']
```

</p></details>

**Comment by iritkatriel** [comment]
CC @ericvsmith @carljm is this something in dataclasses?

**Comment by JelleZijlstra** [comment]
I think this is related to the code in `dataclasses._init_type` (around https://github.com/python/cpython/blob/9a2f2f46caa556eae4c3ac3b45efa85bd91cc807/Lib/dataclasses.py#L746) that looks at `sys.modules` to figure out whether something is an InitVar. `C.__module__` is `__main__` whether or not the module is run through the profiler, but if the profiler is used, `sys.modules["__main__"]` isn't the module defining the dataclass.

I'm not sure what the right solution is. The dataclasses code feels fragile, but it may be the best we can do in the presence of stringified annotations.

**Comment by ericvsmith** [comment]
Agreed on the dataclasses approach being fragile with regard to stringized annotations. But we've not come up with a better approach.

I realize stringized annotations will be around forever, but I'm hoping that if PEP 649 gets finished and merged that we won't care about these cases so much.

**Comment by rhelmot** [comment]
Issue is still present in py 3.11 all the way up through 3.15.0rc1, with both cProfile and profile
